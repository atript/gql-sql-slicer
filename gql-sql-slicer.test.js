const { gqlBuild, merge } = require("./gql-sql-slicer");


describe('new', () => {
  test('distinct', () => {
    expect(gqlBuild(`{
      query{
        device: distinct {
          device
        }
      }
    }
    `).query).toMatchSnapshot();
  })
  test('array on top position', () => {
    const table = [
      {
        "country": "FR"
      },
      {
        "country": "MX"
      },
      {
        "country": "PL"
      },
      {
        "country": "ES"
      },
      {
        "country": "PT"
      },
      {
        "country": "US"
      },
      {
        "country": "FR"
      },
      {
        "country": "IT"
      },
      {
        "country": "GB"
      },
      {
        "country": "DE"
      }
    ]
    const { query, definition } = gqlBuild(`{
      query{
        country: distinct(type: Array) {
            tezt
        }
      }
    }
    `)
    expect(merge(definition, table)).toMatchSnapshot()
  })
})

describe('gqlBuilder', () => {



  test('basic example works', () => {
    expect(gqlBuild(`{
      query: test(brand: Adidas, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
        device {
          date(type: Array) {
            no_baskets
            no_all_baskets
            no_unique_products 
            no_brand_products
            no_uniqie_brand_products 
            total_revenue
            brand_revenue
          }
        }
      }
    }
    `).query).toMatchSnapshot();
  })


  test('metric functions', () => {
    expect(gqlBuild(`{
      query: test(brand: Adidas, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
        device {
          date(type: Array) {
            no_unique_products 
            no_brand_products
            average: divide(a:no_baskets, by:no_all_baskets)
          }
        }
      }
    }
    `).query).toMatchSnapshot();
  })

  test('dimension functions', () => {
    expect(gqlBuild(`{
      query: test(brand: Adidas, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
        device {
          date(type: Array){
            position: aggrAverage(to:no_baskets, by:no_all_baskets) {
              no_baskets 
              no_all_baskets
            }
          }
        }
      }
    }
    `).query).toMatchSnapshot();
  })
  test('group date by month', () => {
    expect(gqlBuild(`{
      query: test(brand: Adidas, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
        device {
          date(type: Array, groupBy:month){
            no_unique_products 
            no_brand_products
          }
        }
      }
    }
    `).query).toMatchSnapshot();
  })
})


describe('merge', () => {

  test('basic example works', () => {
    const table = [
      {
        device: 'mobile',
        date: '2020-01-01T23:00:00.000Z',
        no_baskets: '100',
        no_all_baskets: '101',
        no_unique_products: '102',
        no_brand_products: '103',
        no_uniqie_brand_products: '104',
        total_revenue: '105',
        brand_revenue: '106'
      },
      {
        device: 'mobile',
        date: '2020-01-02T23:00:00.000Z',
        no_baskets: '109',
        no_all_baskets: '107',
        no_unique_products: '108',
        no_brand_products: '110',
        no_uniqie_brand_products: '111',
        total_revenue: '112',
        brand_revenue: '113'
      },
      {
        device: 'desktop',
        date: '2020-01-01T23:00:00.000Z',
        no_baskets: '200',
        no_all_baskets: '201',
        no_unique_products: '202',
        no_brand_products: '203',
        no_uniqie_brand_products: '204',
        total_revenue: '205',
        brand_revenue: '206'
      },
      {
        device: 'desktop',
        date: '2020-01-02T23:00:00.000Z',
        no_baskets: '209',
        no_all_baskets: '207',
        no_unique_products: '208',
        no_brand_products: '210',
        no_uniqie_brand_products: '211',
        total_revenue: '212',
        brand_revenue: '213'
      }
    ]
    const { query, definition } = gqlBuild(`{
      query: test(brand: Adidas, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
        device {
          date(type: Array) {
            no_baskets
            no_all_baskets
            no_unique_products 
            no_brand_products
            no_uniqie_brand_products 
            total_revenue
            brand_revenue
          }
        }
      }
    }
    `);
    expect(merge(definition, table)).toMatchSnapshot()
  })


  test('metric functions', () => {
    const table = [
      {
        device: 'mobile',
        date: '2020-01-01T23:00:00.000Z',
        no_unique_products: '100',
        no_brand_products: '101',
        average: '102'
      },
      {
        device: 'mobile',
        date: '2020-01-02T23:00:00.000Z',
        no_unique_products: '109',
        no_brand_products: '107',
        average: '108'
      },
      {
        device: 'desktop',
        date: '2020-01-01T23:00:00.000Z',
        no_unique_products: '200',
        no_brand_products: '201',
        average: '202'
      },
      {
        device: 'desktop',
        date: '2020-01-02T23:00:00.000Z',
        no_unique_products: '209',
        no_brand_products: '207',
        average: '208'
      }
    ]
    const { query, definition } = gqlBuild(`{
      query: test(brand: Adidas, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
        device {
          date(type: Array) {
            no_unique_products 
            no_brand_products
            average: divide(a:no_baskets, by:no_all_baskets)
          }
        }
      }
    }
    `);
    expect(merge(definition, table)).toMatchSnapshot()
  })

  test('dimension functions', () => {
    const table = [
      {
        device: 'mobile',
        date: '2020-01-01T23:00:00.000Z',
        position_aggrAverage: '100',
        no_baskets: '101',
        no_all_baskets: '102'
      },
      {
        device: 'mobile',
        date: '2020-01-02T23:00:00.000Z',
        position_aggrAverage: '109',
        no_baskets: '107',
        no_all_baskets: '108'
      },
      {
        device: 'desktop',
        date: '2020-01-01T23:00:00.000Z',
        position_aggrAverage: '200',
        no_baskets: '201',
        no_all_baskets: '202'
      },
      {
        device: 'desktop',
        date: '2020-01-02T23:00:00.000Z',
        position_aggrAverage: '209',
        no_baskets: '207',
        no_all_baskets: '208'
      }
    ]
    const { query, definition } = gqlBuild(`{
      query: test(brand: Adidas, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
        device {
          date(type: Array){
            position: aggrAverage(to:no_baskets, by:no_all_baskets) {
              no_baskets 
              no_all_baskets
            }
          }
        }
      }
    }
    `)
    expect(merge(definition, table)).toMatchSnapshot()
  })
  test('date formatting', () => {
    const table = [
      {
        device: 'mobile',
        date: '2020-01-01T23:00:00.000Z',
        position_aggrAverage: '100',
        no_baskets: '101',
        no_all_baskets: '102'
      },
      {
        device: 'mobile',
        date: '2020-01-02T23:00:00.000Z',
        position_aggrAverage: '109',
        no_baskets: '107',
        no_all_baskets: '108'
      },
      {
        device: 'desktop',
        date: '2020-01-01T23:00:00.000Z',
        position_aggrAverage: '200',
        no_baskets: '201',
        no_all_baskets: '202'
      },
      {
        device: 'desktop',
        date: '2020-01-02T23:00:00.000Z',
        position_aggrAverage: '209',
        no_baskets: '207',
        no_all_baskets: '208'
      }
    ]
    const { query, definition } = gqlBuild(`{
      query: test(brand: Adidas, country: US, date_gt: "2020-1-1", date_lt: "2021-7-12") {
        device {
          date(type: Array, format: "Mon yy"){
            position: aggrAverage(to:no_baskets, by:no_all_baskets) {
              no_baskets 
              no_all_baskets
            }
          }
        }
      }
    }
    `)
    expect(merge(definition, table)).toMatchSnapshot()
  })



})

