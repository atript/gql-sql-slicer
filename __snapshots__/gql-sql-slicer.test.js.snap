// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gqlBuilder basic example works 1`] = `"select \\"device\\", \\"date\\", sum(\\"no_baskets\\") as \\"no_baskets\\", sum(\\"no_all_baskets\\") as \\"no_all_baskets\\", sum(\\"no_unique_products\\") as \\"no_unique_products\\", sum(\\"no_brand_products\\") as \\"no_brand_products\\", sum(\\"no_uniqie_brand_products\\") as \\"no_uniqie_brand_products\\", sum(\\"total_revenue\\") as \\"total_revenue\\", sum(\\"brand_revenue\\") as \\"brand_revenue\\" from \\"table\\" where \\"brand\\" = 'Adidas' and \\"country\\" = 'US' and \\"date\\" > '2020-1-1' and \\"date\\" < '2021-7-12' group by \\"device\\", \\"date\\""`;

exports[`gqlBuilder dimension functions 1`] = `"select \\"device\\", \\"date\\", sum(\\"aggrAverage\\")/max(\\"no_all_baskets\\") as position_aggrAverage from (select \\"device\\", \\"date\\", \\"position\\", sum(\\"no_baskets\\") as \\"no_baskets\\", sum(\\"no_all_baskets\\") as \\"no_all_baskets\\", \\"position\\"*sum(\\"no_baskets\\") as \\"aggrAverage\\" from \\"table\\" where \\"brand\\" = 'Adidas' and \\"country\\" = 'US' and \\"date\\" > '2020-1-1' and \\"date\\" < '2021-7-12' group by \\"device\\", \\"date\\", \\"position\\") as \\"middleTable\\" group by \\"device\\", \\"date\\""`;

exports[`gqlBuilder group date by month 1`] = `"select \\"device\\", date_trunc('month', \\"date\\") as \\"date\\", sum(\\"no_unique_products\\") as \\"no_unique_products\\", sum(\\"no_brand_products\\") as \\"no_brand_products\\" from \\"table\\" where \\"brand\\" = 'Adidas' and \\"country\\" = 'US' and \\"date\\" > '2020-1-1' and \\"date\\" < '2021-7-12' group by \\"device\\", date_trunc('month', \\"date\\")"`;

exports[`gqlBuilder metric functions 1`] = `"select \\"device\\", \\"date\\", sum(\\"no_unique_products\\") as \\"no_unique_products\\", sum(\\"no_brand_products\\") as \\"no_brand_products\\", sum(\\"no_baskets\\")/sum(\\"no_all_baskets\\") as \\"average\\" from \\"table\\" where \\"brand\\" = 'Adidas' and \\"country\\" = 'US' and \\"date\\" > '2020-1-1' and \\"date\\" < '2021-7-12' group by \\"device\\", \\"date\\""`;

exports[`merge basic example works 1`] = `
Object {
  "desktop": Array [
    Object {
      "brand_revenue": 206,
      "date": "2020-01-01T23:00:00.000Z",
      "no_all_baskets": 201,
      "no_baskets": 200,
      "no_brand_products": 203,
      "no_uniqie_brand_products": 204,
      "no_unique_products": 202,
      "total_revenue": 205,
    },
    Object {
      "brand_revenue": 213,
      "date": "2020-01-02T23:00:00.000Z",
      "no_all_baskets": 207,
      "no_baskets": 209,
      "no_brand_products": 210,
      "no_uniqie_brand_products": 211,
      "no_unique_products": 208,
      "total_revenue": 212,
    },
  ],
  "mobile": Array [
    Object {
      "brand_revenue": 106,
      "date": "2020-01-01T23:00:00.000Z",
      "no_all_baskets": 101,
      "no_baskets": 100,
      "no_brand_products": 103,
      "no_uniqie_brand_products": 104,
      "no_unique_products": 102,
      "total_revenue": 105,
    },
    Object {
      "brand_revenue": 113,
      "date": "2020-01-02T23:00:00.000Z",
      "no_all_baskets": 107,
      "no_baskets": 109,
      "no_brand_products": 110,
      "no_uniqie_brand_products": 111,
      "no_unique_products": 108,
      "total_revenue": 112,
    },
  ],
}
`;

exports[`merge dimension functions 1`] = `
Object {
  "desktop": Array [
    Object {
      "date": "2020-01-01T23:00:00.000Z",
      "no_all_baskets": 202,
      "no_baskets": 201,
    },
    Object {
      "date": "2020-01-02T23:00:00.000Z",
      "no_all_baskets": 208,
      "no_baskets": 207,
    },
  ],
  "mobile": Array [
    Object {
      "date": "2020-01-01T23:00:00.000Z",
      "no_all_baskets": 102,
      "no_baskets": 101,
    },
    Object {
      "date": "2020-01-02T23:00:00.000Z",
      "no_all_baskets": 108,
      "no_baskets": 107,
    },
  ],
}
`;

exports[`merge metric functions 1`] = `
Object {
  "desktop": Array [
    Object {
      "average": 202,
      "date": "2020-01-01T23:00:00.000Z",
      "no_brand_products": 201,
      "no_unique_products": 200,
    },
    Object {
      "average": 208,
      "date": "2020-01-02T23:00:00.000Z",
      "no_brand_products": 207,
      "no_unique_products": 209,
    },
  ],
  "mobile": Array [
    Object {
      "average": 102,
      "date": "2020-01-01T23:00:00.000Z",
      "no_brand_products": 101,
      "no_unique_products": 100,
    },
    Object {
      "average": 108,
      "date": "2020-01-02T23:00:00.000Z",
      "no_brand_products": 107,
      "no_unique_products": 109,
    },
  ],
}
`;
